### PR TITLE
Multicore ISPC backend (#733)

### PR DESCRIPTION
**Pull request by pema99**
This draft PR adds an extended version of the multicore backend which generates ISPC code (#733).

A few things to point out:
- The output of the multicore backend has been slightly altered, primarily in the form of a few extraneous scoped blocks, and some extra temporary variables here and there.
- We've extended the CI to test the extended backend. Almost of all tests pass, though 3 of them were disabled, primarily due to obscure issues in the ISPC compiler causing miscompilations.
- The maintenance burden of the hand-written ISPC code sections unfortunately seems a bit rough (see uniform.h, scalar.h) - perhaps this code could be generated programmatically in the future?

The extended backend largely just compiles the ImpCode it is given, reusing the multicore backends code where convenient. There are 2 notable tidbits specific to the backend, though:
- Variability of variables throughout the program is inferred at the ImpCode level using a fairly straight forward dependency analysis algorithm.
- The lexical memory block optimization has been altered slightly, and is also repurposed to select which memory can use an, often more performant, pointer-to-vector layout.

### CODE REVIEW

File: src/Futhark/CodeGen/ImpGen/Multicore.hs [hunk 1]
```diff
@@ -40,19 +40,20 @@ gccAtomics = flip lookup cpu
         (Or Int64, Imp.AtomicOr Int64)
       ]
 
--- | Compile the program.
+-- | Compile the program. The boolean indicates whether dynamic scheduling may be used.
```
Comment (athas) [inline src/Futhark/CodeGen/ImpGen/Multicore.hs:43]: Make a type for this.  `data MCOpts = NoDynamicScheduling | YesDynamicScheduling` (or whatever).  [Booleans suck.](http://sboosali.github.io/mirror/boolean-blindness.html)

File: tests/.gitignore [hunk 2]
```diff
@@ -7,3 +7,8 @@
 /**/*.expected
 /**/*.actual
 /**/*.wasm
+/**/*.ispc
+/**/*.h
+/**/*.o
+/**/*.gch
+/**/*.out
```
Comment (athas) [inline tests/.gitignore:14]: Linebreak.

**Review by athas — CHANGES_REQUESTED** [review]
I don't have the time to do a full review right now.  I will be back.

File: rts/c/scalar.h [hunk 3]
```diff
@@ -65,126 +65,303 @@ static inline uint64_t mul64(uint64_t x, uint64_t y) {
   return x * y;
 }
 
+#if ISPC
+
 static inline uint8_t udiv8(uint8_t x, uint8_t y) {
-  return x / y;
+  uint8_t ys = 1;
```
[the remark this answers is not in the document]
Comment (pema99) [reply rts/c/scalar.h:71]: Every place it is used, or just once? It's unfortunately fairly common 😬

[the remark this answers is not in the document]
Comment (pema99) [reply rts/c/scalar.h:75]: added it once, at least

File: tests/.gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (pema99) [reply tests/.gitignore]: [81b1f90](https://github.com/diku-dk/futhark/pull/1677/commits/81b1f90aa52481c0f434e17ae8c17c857905711c)

File: src/Futhark/CodeGen/ImpGen/Multicore.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pema99) [reply src/Futhark/CodeGen/ImpGen/Multicore.hs]: [1a062e0](https://github.com/diku-dk/futhark/pull/1677/commits/1a062e0414a4ebe922db6ac204225288b0ecda2d)

**Comment by athas** [comment]
To shrink this PR, I am going to morally move independently useful parts of it into `master` (e.g. the change to `Copy`) so the essence becomes clearer.

**Comment by athas** [comment]
Is there any meaningful advantage to using the header file generated by ISPC, rather than just constructing one yourselves?  You seem to have a lot of trouble with it, and in practice you assume it to have very specific contents anyway.  I don't think there is any advantage to having ISPC generate it.

**Comment by pema99** [comment]
> Is there any meaningful advantage to using the header file generated by ISPC, rather than just constructing one yourselves? You seem to have a lot of trouble with it, and in practice you assume it to have very specific contents anyway. I don't think there is any advantage to having ISPC generate it.

Hmm, I don't think we ever really considered the possibility. I don't see any fundamental reason we couldn't just generate it ourselves. The main thing making the auto-generated ones annoying to deal with is that you have little control over when structs definitions are exported. I opened an issue on it which was acknowledged (https://github.com/ispc/ispc/issues/2277).

**Comment by athas** [comment]
Generating your own header file would solve that problem (and perhaps allow other things to be done more easily in the future).  It would also avoid future related issues.  It also makes it easier to use the backend in a library setting (right now it's a bit mysterious that you have to compile the .ispc file to get the .h file).

**Comment by pema99** [comment]
> Generating your own header file would solve that problem (and perhaps allow other things to be done more easily in the future).  It would also avoid future related issues.  It also makes it easier to use the backend in a library setting (right now it's a bit mysterious that you have to compile the .ispc file to get the .h file).

I agree. We can look into generating these ourselves tomorrow 👍 will update here if we happen to stumble upon some unforeseen issue preventing this

**Comment by athas** [comment]
There is in fact no need to generate a header file at all.  Just insert the necessary prototypes directly into the `.c` file.

**Comment by pema99** [comment]
We've made change you requested. All the tests still seem to pass :D

**Comment by athas** [comment]
And you even got rid of stack.yaml as well.

**Comment by pema99** [comment]
FWIW, the command currently is `futhark ispc-multicore`, not `futhark ispc`. Would you like to change that? It was more or less just pattern matched from `wasm-multicore`.

**Comment by athas** [comment]
Yes, please change that.

**Comment by athas** [comment]
When compiling to a library, it is impractical that `foo.fut` gives rise to both `foo.c` and `foo.ispc`, because the natural compilation target for both of these files is `foo.o`.  I suggest the ISPC program file is given the extension `.kernels.ispc` instead.

**Comment by pema99** [comment]
Missed a spot in that first commit. CI should pass again now. I made both changes you requested

**Review by athas — CHANGES_REQUESTED** [review]

File: tests/soacs/reduce8.fut [hunk 4]
```diff
@@ -0,0 +1,9 @@
+-- This test tests an explicitly commutative reduction over scalars.
```
[the remark this answers is not in the document]
Comment (pema99) [reply tests/soacs/reduce8.fut:1]: Not really, but I also don't think it hurts to have. Do you want it removed?

File: src/Futhark/CodeGen/Backends/GenericC.hs [hunk 5]
```diff
@@ -651,7 +677,7 @@ defineMemorySpace space = do
   free <- collect $ freeRawMem [C.cexp|block->mem|] space [C.cexp|desc|]
   ctx_ty <- contextType
   let unrefdef =
-        [C.cedecl|static int $id:(fatMemUnRef space) ($ty:ctx_ty *ctx, $ty:mty *block, const char *desc) {
```
[the remark this answers is not in the document]
Comment (pema99) [reply src/Futhark/CodeGen/Backends/GenericC.hs:654]: How so? Naming conflicts with other compilation units?

[the remark this answers is not in the document]
Comment (athas) [reply src/Futhark/CodeGen/Backends/GenericC.hs:654]: Yes.

File: tests/soacs/reduce8.fut (hunk 4 above)
[the remark this answers is not in the document]
Comment (athas) [reply tests/soacs/reduce8.fut:1]: Yes.

File: src/Futhark/CodeGen/Backends/GenericC.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pema99) [reply src/Futhark/CodeGen/Backends/GenericC.hs:654]: I see :grimacing: is it fine to keep like this for now, or do you have any simple solutions in mind?

File: src/Futhark/CodeGen/ImpGen/Multicore/SegScan.hs [hunk 6]
```diff
@@ -54,79 +54,184 @@ nonsegmentedScan ::
 nonsegmentedScan pat space scan_ops kbody nsubtasks = do
   emit $ Imp.DebugPrint "nonsegmented segScan" Nothing
   collect $ do
-    scanStage1 pat space scan_ops kbody
+    -- Are we working with nested arrays
+    let dims = map (shapeDims . segBinOpShape) scan_ops
+    let isScalar x = case x of MemPrim _ -> True; _ -> False
```
Comment (athas) [inline src/Futhark/CodeGen/ImpGen/Multicore/SegScan.hs:59]: primType . typeOf

File: src/Futhark/CodeGen/ImpGen/Multicore/SegScan.hs [hunk 7]
```diff
@@ -54,79 +54,184 @@ nonsegmentedScan ::
...
+    -- Are we only working on scalars
+    let scalars = all (all (isScalar . paramDec) . (lambdaParams . segBinOpLambda)) scan_ops && all (== []) dims
+    -- Do we have nested vector operations
+    let vectorize = [] `notElem` dims
+
+    let param_types = concatMap (map paramType . (lambdaParams . segBinOpLambda)) scan_ops
+    let no_array_param = null [x | x@(Array {}) <- param_types]
```
Comment (athas) [inline src/Futhark/CodeGen/ImpGen/Multicore/SegScan.hs:66]: `all primType param_types`

File: src/Futhark/CodeGen/ImpGen/Multicore/Base.hs [hunk 8]
```diff
@@ -226,17 +235,19 @@ extractAllocations segop_code = f segop_code
     f code =
       (mempty, code)
 
+data ChunkLoopVectorization = Vectorized | Scalar
```
Comment (athas) [inline src/Futhark/CodeGen/ImpGen/Multicore/Base.hs:238]: Please add a comment indicating the purpose of this.

File: src/Futhark/CodeGen/ImpGen/Multicore/SegHist.hs [hunk 9]
```diff
@@ -211,35 +227,57 @@ subHistogram pat space histops num_histos kbody = do
...
+                    dest_shape' = map toInt64Exp $ shapeDims dest_shape
+                    acc_params' = (lambdaParams . histOp) histop'
+                    vs_params' = takeLast (length vs') $ lambdaParams $ histOp histop'
+
+                generateUniformizeLoop $ \j ->
+                  sComment "perform updates" $ do
+                    -- Create new set of uniform buckets
+                    -- That is extract each bucket from a SIMD vector lane
+                    extract_buckets <- mapM (dPrim "extract_bucket" . (primExpType . untyped)) bucket'
+                    mapM_
```
Comment (athas) [inline src/Futhark/CodeGen/ImpGen/Multicore/SegHist.hs:259]: Use `forM_`.

**Review by athas — CHANGES_REQUESTED** [review]

File: src/Futhark/CodeGen/ImpGen/Multicore/SegScan.hs [hunk 10]
```diff
@@ -54,79 +54,183 @@ nonsegmentedScan ::
 nonsegmentedScan pat space scan_ops kbody nsubtasks = do
   emit $ Imp.DebugPrint "nonsegmented segScan" Nothing
   collect $ do
-    scanStage1 pat space scan_ops kbody
+    -- Are we working with nested arrays
+    let dims = map (shapeDims . segBinOpShape) scan_ops
+    -- Are we only working on scalars
+    let scalars = all (all (primType . typeOf . paramDec) . (lambdaParams . segBinOpLambda)) scan_ops && all (== []) dims
```
Comment (athas) [inline src/Futhark/CodeGen/ImpGen/Multicore/SegScan.hs:60]: `(== [])` is better written `null`.

File: src/Futhark/CodeGen/ImpGen/Multicore/SegHist.hs [hunk 11]
```diff
@@ -211,35 +227,55 @@ subHistogram pat space histops num_histos kbody = do
...
+                    acc_params' = (lambdaParams . histOp) histop'
+                    vs_params' = takeLast (length vs') $ lambdaParams $ histOp histop'
+
+                generateUniformizeLoop $ \j ->
+                  sComment "perform updates" $ do
+                    -- Create new set of uniform buckets
+                    -- That is extract each bucket from a SIMD vector lane
+                    extract_buckets <- mapM (dPrim "extract_bucket" . (primExpType . untyped)) bucket'
+                    forM_ (zip extract_buckets bucket') $
+                      \(x, y) ->
```
Comment (athas) [inline src/Futhark/CodeGen/ImpGen/Multicore/SegHist.hs:260]: Put this on the same line as the `$`.

**Review by athas — CHANGES_REQUESTED** [review]

**Comment by athas** [comment]
Is there a reason not to merge this now?  Is anything missing that is realistic to fix?

**Review by athas — APPROVED** [review]

**Comment by olivbak** [comment]
> Is there a reason not to merge this now? Is anything missing that is realistic to fix?

We are currently investigating optimizations for small ISPC kernels, discussed in Question 1 we sent you last week. Otherwise no.

**Comment by athas** [comment]
Well there are also a bunch of the benchmarks that don't validate, but I think that goes into the category of "not likely to be fixed".  Maybe even ISPC compiler bugs.

**Comment by olivbak** [comment]
> Well there are also a bunch of the benchmarks that don't validate, but I think that goes into the category of "not likely to be fixed". Maybe even ISPC compiler bugs.

By validate, do you refer to worse performance compared to the MC backend? If you decide to merge it, we can just open another PR with optimizations if we happen to fix them

**Comment by athas** [comment]
No, I mean that they don't give the right results:

```
$ futhark bench --backend=ispc futhark-benchmarks/jgf/series/series.fut
Compiling futhark-benchmarks/jgf/series/series.fut...
Reporting arithmetic mean runtime of at least 10 runs for each dataset (min 0.5s).
More runs automatically performed for up to 300s to ensure accurate measurement.

futhark-benchmarks/jgf/series/series.fut (no tuning file):
data/10000.in:
futhark-benchmarks/jgf/series/series.fut.actual and futhark-benchmarks/jgf/series/series.fut.expected do not match:
Value #1 index [3000]: expected 0.0, got 5.618383934408205e-3
...and 13 other mismatches.
data/100000.in:
futhark-benchmarks/jgf/series/series.fut.actual and futhark-benchmarks/jgf/series/series.fut.expected do not match:
Value #0 index [12999]: expected 1.116105, got 1.1043930678879241
...and 1640 other mismatches.
```

**Comment by pema99** [comment]
@athas 
> No, I mean that they don't give the right results:
Yeah, we noticed that as well. Most of the ones I found weren't too far off the right result, though.

> Well there are also a bunch of the benchmarks that don't validate, but I think that goes into the category of "not likely to be fixed". Maybe even ISPC compiler bugs.
I pretty much agree with this assesment.

From my POV the PR is pretty much ready to go. The last changes we wanted to include are on this branch https://github.com/diku-dk/futhark/commits/ispc-from-to 
Which includes the changes to `ForEach` that we discussed. If those look fine to you, I can merge it in, and which I point I don't think we'll have anything else to add.

**Comment by athas** [comment]
Looks fine to me.

**Comment by pema99** [comment]
Alright, with that merged, I think we're satisfied.

**Comment by athas** [comment]
```
$ futhark datacmp series.fut.actual series.fut.expected
Value #1 index [3000]: expected 0.0, got 5.618383934408205e-3
Value #1 index [4000]: expected 0.0, got 7.491177291036956e-3
Value #1 index [5000]: expected 0.0, got 9.36396954380273e-3
Value #1 index [6000]: expected 0.0, got 1.1236760416497281e-2
Value #1 index [6999]: expected 1.881969, got 1.886173077484973
Value #1 index [7000]: expected 0.0, got 1.3109549632837015e-2
Value #1 index [7001]: expected -1.881969, got -1.8777470265875036
Value #1 index [7999]: expected 1.881969, got 1.8867722032380607
Value #1 index [8000]: expected 0.0, got 1.4982336917322711e-2
Value #1 index [8001]: expected -1.881969, got -1.8771424369015954
Value #1 index [8999]: expected 1.881969, got 1.8873709635069034
Value #1 index [9000]: expected 0.0, got 1.6855121993659482e-2
Value #1 index [9001]: expected -1.881969, got -1.8765374841782205
Value #1 index [9999]: expected 1.881969, got 1.887969358138663
```

If this is due to floating-point jitter, then I'll eat my hat.  (Bought at [Petitgas](https://petitgas.dk/); support your local shops.)

**Comment by pema99** [comment]
I have no clue what is up with those test failures on the last merge o.o

**Comment by pema99** [comment]
@athas I think I found what is causing the discrepancy in the `series` benchmark. Floating point literals in ISPC are apparently always 32 bit precision unless specified otherwise. The C and ISPC programs for that benchmark already diverge at one of the first calculations, which looks like so:
```
3.141592653589793 * (double)(x)
```
Where `x=3000`. ISPC returns `9424.778223` while C returns `9424.777961`, because the pi literal is being treated as 32 bit in ISPC. Appending the literal with a `d` to indicate double causes the 2 programs to produce the same output.

**Comment by pema99** [comment]
Actually.. Had to revert that last commit. The idea is right, and makes the benchmark succeed but causes some other issues...

**Comment by pema99** [comment]
Now it should be sound, I think

**Comment by athas** [comment]
Have you noticed how long [poseidon](https://github.com/diku-dk/futhark-benchmarks/tree/master/misc/poseidon) takes to compile?

**Comment by pema99** [comment]
> Have you noticed how long [poseidon](https://github.com/diku-dk/futhark-benchmarks/tree/master/misc/poseidon) takes to compile?

I did notice it a while back, but never investigated it, and later forgot about it. I'm not sure what takes so long. I just tried compiling it on my machine, and it still wasn't done after 10 minutes. FWIW I see that same behavior using the multicore backend on `master` without any of our changes.

**Comment by pema99** [comment]
How do you usually go about profiling the compiler, btw?

**Comment by athas** [comment]
Pema Malling ***@***.***> writes:

> How do you usually go about profiling the compiler, btw?

cabal configure --enable-profiling

-- 
\  Troels
/\ Henriksen

**Comment by athas** [comment]
Here's a profiling report: ttp://sigkill.dk/junk/futhark.prof

While the pipeline is known to be overall slow on this program, the ISPC codegen is dramatically slow.  The profiling report is actually from a run I aborted because it took so long (but you can still see that `compileOp` calls `namesFromList` and for some reason that takes forever.

**Comment by pema99** [comment]
I hadn't fully realized this, but in hindsight it's pretty obvious. Gonna play around with the code a bit

**Comment by pema99** [comment]
Well, a few observations.

It's definitely the variability analysis causing the slowdown. Disabling it causes compilation to take roughly as long as the multicore backend.

I thought we might be able to just traverse the 'sequentialized' version of each outermost SegOp for variability analysis and use the results in the rest of the code generation for that SegOp, but it won't work. First of all, the names of ImpCode variables don't necessarily match in the 2 versions, and secondly, the desired variability for the same variable may actually be different in the 2 versions. I don't think we can work around that without some rethinking that we probably won't have the time for. Maybe we should just treat this as future work?

Also, just to clarify, when I said I thought the code was quadratic, I meant in terms of the level of nesting of SegOps. That's still not great, but at least it's not in terms of the code length. If the sequentialized versions of each SegOp didn't exist, there wouldn't be an issue.

Lastly, I should say that my analysis is sort of speculative. I know that variability analysis is causing the slowdown, and the bad scaling is probably part of that, but I don't know to which extent, and haven't looked much at how much nesting is present in the poseidon program. For all I know, it could just be a small portion, and some other slow operation in that codepath is the real culprit. I haven't looked at the profiler output closely enough

I assume the 27% timing you see at the top of the profiler output you attached is an aggregate? Is there a way to easily locate the hottest path?

**Comment by athas** [comment]
You run with `+RTS -p` to produce the .prof file, and then there are various tools you can use to analyse the raw results, such as [profiterole](https://hackage.haskell.org/package/profiterole) (or maybe something newer; I haven't kept up with the tooling).

**Comment by pema99** [comment]
The changes in that last commit sped up the variability analysis by almost 5x on my machine. Testing with a slightly trimmed down version of poseidon, the ISPC backend now takes a bit over double the total compilation time of the multicore backend for me.

**Comment by pema99** [comment]
@athas About dynamic scheduling being disabled on the ISPC backend: We thought a bit about what we discussed today, and as an experiment tried reenabling dynamic scheduling, but hardcoding the minimum chunk size to 8 instead of the current 1. It yielded quite a large speedup on the mandelbrot benchmark (about x1.8). It also sped up the 2 raytracers I tried slightly.

With that realization, I feel like we should maybe reenable dynamic scheduling, and let the minimum chunk size be the gang size of the target instruction set.. What do you think?

**Comment by athas** [comment]
Makes sense to me.

I intend to merge this PR once [this one](https://github.com/NixOS/nixpkgs/pull/173136) has gone in.  I'd rather make sure we have all testing dependencies available via Nix.  (Although I am yet undecided on whether to run the ISPC benchmarks automatically because this would *significantly* increase benchmarking time.)

**Comment by athas** [comment]
No.  I'll get around to it soon.

**Comment by athas** [comment]
I'll change CI to use `ispc` from Nix tomorrow, if someone doesn't do it before me.  I think that's the last missing piece.

**Pull request merged**